### ISSUE TITLE
Poor performance of table vectorizer at serving time

### ISSUE THREAD

**Issue by TTRh** [skrub-data/skrub#1372]
### Describe the bug

Hi,

(First i'm not sure if this should be classified as bug or discussion, feel free to change the category :) )

I'll like to report an issue i have using TableVectorizer, I'm using it to handle a large number of heterogeneous text & numeric features (short text are identifier like countries, cities, address, names etc ...).
I'm using it part of a larger scikit learn pipeline where:
* final estimator is an hist gradient boosting classifier
* i'm passing the output of table vectorizer to the final estimator along with other features i join with scikit learn FeatureUnion. 

Problem i have is that at serving time i need to serve prediction not by batch but sample by sample and when i introduced table vectorizer to the model serving time went from ~150 milisec to ~2minutes to do a prediction for 1 sample which is definitely not usable as it in production.

I identified that the latency comes in part from the gap encoder based on the profile log but i don't know if this is kind of normal or there is an issue behind.

cProfile log file when calling the model.predict_proba method with only 1 sample

```
import cProfile
cProfile.run("model.predict_proba(X[:1])[:, 1]")
```

[cprofile.log](https://github.com/user-attachments/files/20067165/cprofile.log)

I'm not able to provide a simple code to reproduce the issue, but feel free to tell me what i could i provide and i'll try to share what i can.

Number of input raw feature for the table vectorizer ~225
And distribution of `kind_to_columns` : 
```
{'numeric': 138,
 'datetime': 0,
 'low_cardinality': 41,
 'high_cardinality': 39,
 'specific': 0}
```

### Steps/Code to Reproduce

NA

### Expected Results

NA

### Actual Results

NA

### Versions

```shell
System:
    python: 3.10.4 (main, Jan 27 2025, 17:17:32) [GCC 11.4.0]
executable: /home/john.doe/.cache/pypoetry/virtualenvs/my-project/bin/python
   machine: Linux-6.8.0-58-generic-x86_64-with-glibc2.35

Python dependencies:
      sklearn: 1.2.2
          pip: 23.2.1
   setuptools: 68.1.2
        numpy: 1.25.2
        scipy: 1.11.2
       Cython: None
       pandas: 2.0.3
   matplotlib: 3.7.2
       joblib: 1.3.2
threadpoolctl: 3.2.0

Built with OpenMP: True

threadpoolctl info:
       user_api: blas
   internal_api: openblas
    num_threads: 14
         prefix: libopenblas
       filepath: /home/john.doe/.cache/pypoetry/virtualenvs/my-project/lib/python3.10/site-packages/numpy.libs/libopenblas64_p-r0-5007b62f.3.23.dev.so
        version: 0.3.23.dev
threading_layer: pthreads
   architecture: Haswell

       user_api: openmp
   internal_api: openmp
    num_threads: 14
         prefix: libgomp
       filepath: /home/john.doe/.cache/pypoetry/virtualenvs/my-project/lib/python3.10/site-packages/scikit_learn.libs/libgomp-a34b3233.so.1.0.0
        version: None

       user_api: blas
   internal_api: openblas
    num_threads: 14
         prefix: libopenblas
       filepath: /home/john.doe/.cache/pypoetry/virtualenvs/my-project/lib/python3.10/site-packages/scipy.libs/libopenblasp-r0-23e5df77.3.21.dev.so
        version: 0.3.21.dev
threading_layer: pthreads
   architecture: Haswell
0.5.1
```

**Comment by TTRh** [comment]
Yes, first we used the default of table vectorizer then we did a comparison trying different options (time & metrics are super specific to our problem) but here was the result if this can be useful: 

| Experiment              | fit time | predict time | precision (%) | recall (%) |
| :---------------- | :------: | :----: | :----: | :----: |
| table vectorizer default        |   2h4min   | 20min | 23.39 | 98.21 |
| gap encoder with reduced dimension    |  1h46min   | 9min | 21.74 | 98.41 |
| gap encoder with hashing |   1h25min   | 30min | 23.10 | 98.11 |
| string encoder |   1h41min   | 5min | 23.11 | 98.20 |
| hashing encoder  |   59min   | 4min | 22.65 | 98.13 |

So we went with the string encoder for high_cardinality of table vectorizer as it was the best tradeoff.
Serving performance is not yet what we expect though (takes around 1 sec to give a prediction for 1 sample for our current model described above) and we have hard time to make our serving service scale with the load (compared to pure sklearn models).

New profile log file with string encoder if this can be useful: 

[cprofile_string_encoder.log](https://github.com/user-attachments/files/20067643/cprofile_string_encoder.log)

**Comment by Vincent-Maladiere** [comment]
Hey @TTRh, thank you for your feedback. These benchmarks are extremely useful to us!
I see you have 39 high cardinality columns. Do you have any IDs among them? I'm asking because IDs seldom contain meaningful features to extract for your downstream task, and encoding them is a big memory and computation strain. In that case, you should exclude them before running the table vectorizer.

**Comment by TTRh** [comment]
thank your for your feedback!

> I see you have 39 high cardinality columns. Do you have any IDs among them? 
not pure ids but similar i guess, i have for instance features like: user agent / browser name & version / gpu name / os name & version / internet service provider name ... 

> To get speed, one option would be to use scikit-learn TargetTransformer as a high-cardinality encoder. Do you want to try this? 

I can do a test with the target transformer yes 👍

**Comment by GaelVaroquaux** [comment]
We're discussing this issue at our skrub weekly meeting. Just to check: you have "2mn" prediction time for one sample?

We find this surprisingly long. The only explanation that we find is that you could have very very long texts. Is that the case?

What's your baseline without skrub? Your "standard" way of working?

**Comment by TTRh** [comment]
> Just to check: you have "2mn" prediction time for one sample?
 
yes, i shared the profiling log in the first message: i got "149.833 seconds" there is a small overhead passing it to cProfile.run but testing it (on my laptop ~32Gb / cpu intel ultra 7 165U w/ 14 cpu) i get in average 2min for 1 sample prediction. Same features changing gap encoder by string encoder the profiling drop at around 1-2 sec for 1 sample prediction still.

>  The only explanation that we find is that you could have very very long texts. Is that the case?

I don't have super long text on the features i pass to table vectorizer, i can double check and give you the correct number but my gess is that max text length woulnd't go over 100 chars length.

> What's your baseline without skrub?

| Experiment              | fit time | prediction time (for test set) | precision (%) | recall (%) |
| :---------------- | :----: | :----: | :----: | :----: |
| base model without high cardinality features  |  54min  | 3min | 97.39 | 11.92 |
| base model with high cardinality features without skrub | 1h45min  | 3min | 97.73 | 17.05 |
| model with skrub with table vectorizer default | 2h4min | 20min |	98.21 | 23.39 |

When we first saw this we were quite happy with the preformance of skrub as our recall got a huge boost with it ...

**Comment by Vincent-Maladiere** [comment]
Thanks for the update @TTRh! Could you add entries in your table with:
- model with `skrub.TableVectorizer(high_cardinality=skrub.StringEncoder())`
- model with `skrub.TableVectorizer(high_cardinality=sklearn.preprocessing.TargetEncoder())`
This will help us understand how these encoders behave with your dataset :)

**Comment by TTRh** [comment]
Yes,let me recap all experiments i did so far:
 
| Experiment              | fit time | prediction time (for test set) | recall (%) | precision (%) | time for 1 sample prediction | profiling |
| :---------------- | :----: | :----: | :----: | :----: | :----: | :----: |
| base model <br> * without high cardinality features  |  54min  | 3min | 11.92 | 97.39 | 
| model <br> * with high cardinality features <br> * without skrub | 1h45min  | 3min | 17.05 |97.73 | 
| model with <br> `skrub.TableVectorizer(high_cardinality=skrub.GapEncoder())` | 2h4min | 20min |	23.39 |98.21 | ~2min | [cprofile.log](https://github.com/user-attachments/files/20187135/cprofile.log) |
| model with <br> `skrub.TableVectorizer(cardinality_threshold=60, high_cardinality=skrub.GapEncoder(n_components=10))`    |  1h46min   | 9min | 21.74 | 98.41 |
| model with <br>  `skrub.TableVectorizer(high_cardinality=skrub.GapEncoder(hashing=True))` |   1h25min   | 30min | 23.10 | 98.11 |
| model with <br> `skrub.TableVectorizer(high_cardinality=skrub.StringEncoder())` |   1h41min   | 5min | 23.11 | 98.20 | ~2sec | [cprofile.log](https://github.com/user-attachments/files/20187145/cprofile_string_encoder.log) |
| model with <br> `skrub.TableVectorizer(high_cardinality=skrub.MinHashEncoder())`   |   59min   | 4min | 22.65 | 98.13 |
| model with <br> `skrub.TableVectorizer(high_cardinality=sklearn.preprocessing.TargetEncoder())`  |  1h2min  | 3min | 36.77 | 97.12 |

I'll continue to update this table with the missing values.
EDIT: updated with TargetEncoder

**Comment by Vincent-Maladiere** [comment]
Nice improvement on the prediction time with the StringEncoder instead of the GapEncoder!

**Comment by TTRh** [comment]
Hello :) Sorry for the delay (had to upgrade project sklearn version + holidays in between 🙈).
Got finally the experiment with TargetEncoder and it definitely worth it, got my best score so far with this one. I updated the table above with this.
Thanks a lot for all the suggestions & help.

**Comment by GaelVaroquaux** [comment]
> Got finally the experiment with TargetEncoder and it definitely worth it, got my best score so far with this one. I updated the table above with this.

That is super useful feedback.

Ok, I consider that we have found the solution to the present issue: **we need to document better the tradeoffs and make it easier for people to find the TargetEncoder**

**Comment by TTRh** [comment]
Hello guys, i've been trying to dig a bit on the gap encoder slowness, i've a small snippet code which (i think) reproduce more or less the issue i have. I'm sharing this here: 

```python
from sklearn.datasets import fetch_20newsgroups
from sklearn.model_selection import train_test_split
from skrub import GapEncoder
import pandas as pd
import random
import time

# build dataset
dataset = fetch_20newsgroups(subset='all')
X = dataset.data
y = dataset.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
print(f"Size train: {len(X_train)}, size test: {len(X_test)}")

# fit
start = time.time()
encoder = GapEncoder(n_components=30)
encoder.fit(pd.Series(X_train), y_train)
end = time.time()
print(f"Fit time = {(end - start):.2f} sec")

# predict test set
start = time.time()
encoder.transform(pd.Series(X_test))
end = time.time()
print(f"Predict time (X_test) ={(end - start):.2f} sec")

# predict on 1 sample
start = time.time()
encoder.transform(pd.Series(X_test).head(1))
end = time.time()
print(f"Predict time (1 sample) = {(end - start):.2f} sec")
```

```
# result for test_size = 0.3
Size train: 13192, size test: 5654
Fit time = 101.59 sec
Predict time (X_test) = 200.76 sec
Predict time (1 sample) = 100.31 sec

# result for test_size = 0.7 (so smaller X_train)
Size train: 7538, size test: 11308
Fit time = 74.16 sec
Predict time (X_test) = 182.68 sec
Predict time (1 sample) = 38.90 sec
```

I did not dig deep enough on the gap encoder implementation but looks like bigger is the train set, slower is the prediction on the one sample. I don't know if this is expected ? At least it explained what i'm seeing on my own dataset.

I also realized it might be related to how "clean" the input text is, for instance testing with even bigger dataset (x5 bigger) with faker i don't get the same issue at all:

```python
from sklearn.model_selection import train_test_split
from skrub import GapEncoder
import pandas as pd
from faker import Faker
import random
import time

fake = Faker()

# build dataset
size_dataset = 100000
X = [fake.text(700) for _ in range(size_dataset)]
y = random.choices([0, 1], k=size_dataset)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
print(f"Size train: {len(X_train)}, size test: {len(X_test)}")

# fit
start = time.time()
encoder = GapEncoder(n_components=30)
encoder.fit(pd.Series(X_train), y_train)
end = time.time()
print(f"Fit time = {(end - start):.2f} sec")

# predict test set
start = time.time()
encoder.transform(pd.Series(X_test))
end = time.time()
print(f"Predict time (X_test) = {(end - start):.2f} sec")

# predict on 1 sample
start = time.time()
encoder.transform(pd.Series(X_test).head(1))
end = time.time()
print(f"Predict time (1 sample) = {(end - start):.2f} sec")
```

I get

```
Size train: 70000, size test: 30000
Fit time = 71.18 sec
Predict time (X_test) = 101.45 sec
Predict time (1 sample) = 4.90 sec
```

I'm thinking that because fake.text() generates cleaner text (real world, no typo, small amount of numerical values & punctuation etc ...) it's much easier for the similarity computation behind ? 

I also noticed on the newsgroup dataset strange burst of memory usage on predict both for the full X_test or one sample. Did not notice the same thing for faker dataset.

Could it make sense to add some pre-processing / cleaning step before gap encoding or it would reduce the added value of it ?

Curious to have your feedback on that :)
